File: algorithms/javaalgo/src/test/java/programmers/level2/nnumbergame/NNumberGameTest.java
package programmers.level2.nnumbergame;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import org.junit.jupiter.api.Test;

public class NNumberGameTest {

    @Test
    void test() {
        int n1 = 2;
        int t1 = 4;
        int m1 = 2;
        int p1 = 1;

        int n2 = 16;
        int t2 = 16;
        int m2 = 2;
        int p2 = 1;

        int n3 = 16;
        int t3 = 16;
        int m3 = 2;
        int p3 = 2;

        assertAll(
            () -> assertThat(solution(n1, t1, m1, p1)).isEqualTo("0111"),
            () -> assertThat(solution(n2, t2, m2, p2)).isEqualTo("02468ACE11111111"),
            () -> assertThat(solution(n3, t3, m3, p3)).isEqualTo("13579BDF01234567")
        );
    }

    public String solution(int n, int t, int m, int p) {
        StringBuilder answer = new StringBuilder();
        StringBuilder sb = new StringBuilder();

        for (int i = 0; sb.length() < t * m; i++) {
            sb.append(Integer.toString(i, n).toUpperCase());
        }

        for (int i = p - 1; answer.length() < t; i += m) {
            answer.append(sb.charAt(i));
        }

        return answer.toString();
    }
}

File: algorithms/javascriptalgo/programmers/level2/n진수_게임/nNumberGame.test.js
test("test", () => {
  let n1 = 2;
  let t1 = 4;
  let m1 = 2;
  let p1 = 1;

  let n2 = 16;
  let t2 = 16;
  let m2 = 2;
  let p2 = 1;

  let n3 = 16;
  let t3 = 16;
  let m3 = 2;
  let p3 = 2;

  expect(solution(n1, t1, m1, p1)).toBe("0111");
  expect(solution(n2, t2, m2, p2)).toBe("02468ACE11111111");
  expect(solution(n3, t3, m3, p3)).toBe("13579BDF01234567");
});

function solution(n, t, m, p) {
  let answer = "";
  let nNumber = "";
  for (let i = 0; i < t * m; i++) {
    nNumber += i.toString(n).toUpperCase();
  }
  for (let i = p - 1; i < t * m; i += m) {
    answer += nNumber[i];
  }

  return answer;
}
